File: tests/__tests__/api/upload_test.js
const frisby = require('frisby');
const sha1 = require('js-sha1');
const qs = require('qs');
var path = require('path');
var fs = require('fs');

const BASE_URL = "http://je.test.com:2330/project/public";


describe('图片上传测试', function() {


    var request_header = {
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded', //
        }
    };
    frisby.globalSetup({
        request: request_header
    });

    // getNonce 测试
    var nonce = "";

    it('获取 nonce', function() {
        return frisby.post(BASE_URL + '/auth/nonce')
            .expect('json', 'ret', 200)
            .expect('jsonTypes', 'data.nonce', frisby.Joi.string())
            .then(function(res) {
                console.log("nonce: ", res._json.data.nonce)
                nonce = res._json.data.nonce;
            });
    });

    // 登录测试

    var token = "";
    it('用户登录', async function() {

        var loginBody = qs.stringify({
            email: "i@pluvet.com",
            nonce: nonce,
            sign: sha1(nonce + "i@pluvet.com" + sha1("moeje" + "apitesting")),
        });

        return frisby.post(BASE_URL + '/auth/login/email', { body: loginBody })
            .expect('json', 'ret', 200)
            .expect('jsonTypes', 'data.token', frisby.Joi.string())
            .then(function(res) {
                console.log("token: ", res._json.data.token)
                token = res._json.data.token;
            });
    });

    it('上传一个txt文件, 期待返回 MIME 错误', function() {
        var imagePath = path.resolve(__dirname, 'file.txt');
        var form = frisby.formData();

        form.append('file', fs.createReadStream(imagePath), {
            knownLength: fs.statSync(imagePath).size
        });
        frisby // Post image
            .setup({
                request: {
                    headers: {
                        'content-type': 'multipart/form-data; boundary=' + form.getBoundary(),
                        'content-length': form.getLengthSync()
                    }
                }
            })
            .post(BASE_URL + '/upload/image', {
                body: form
            })
            .expect('status', 200)
            .expect('json', 'ret', 23333)
            .expect('json', 'msg', '非法请求：参数file应该为：image/jpeg/image/png/image/gif/image/webp，但现在file = text/plain')

    });
    it('上传一个png文件', function() {
        var imagePath = path.resolve(__dirname, 'google.png');
        var form = frisby.formData();

        form.append('file', fs.createReadStream(imagePath), {
            knownLength: fs.statSync(imagePath).size
        });
        frisby // Post image
            .setup({
                request: {
                    headers: {
                        'content-type': 'multipart/form-data; boundary=' + form.getBoundary(),
                        'content-length': form.getLengthSync()
                    }
                }
            })
            .post(BASE_URL + '/upload/image', {
                body: form
            })
            .expect('json', 'ret', 200)

    });
});

File: tests/py/upload_test.py
import time
import hashlib
import requests
import util.log as log
import json
import jsbeautifier
from pprint import pprint
from stopwatch import Stopwatch


import os
import sys

testNum = 0


def getNum():
    global testNum
    testNum += 1
    return ' #' + str(testNum) + ' '


def sha1(str):
    return hashlib.sha1(str.encode("utf-8")).hexdigest()


log.prepare()

baseUrl = 'http://127.0.0.1:2330/project/public'

log.info(" ================================ 文件上传测试 ================================ ")
print()

stopwatch = Stopwatch()
stopwatch.start()

try:

    res = requests.post(baseUrl + "/auth/nonce",
                        data={}, allow_redirects=False)
    jsonRaw = jsbeautifier.beautify(res.text)
    #log.response(jsonRaw, 1)
    jsonObj = json.loads(jsonRaw)
    jsonPass = True
    email = 'i@pluvet.com'
    password = 'apitesting'
    nonce = jsonObj['data']['nonce']
    sign = sha1(nonce + email + sha1("moeje" + password))
    log.info("sign: " + sign)
    res = requests.post(baseUrl + "/auth/login/email",
                        data={
                            'email': email,
                            'nonce': nonce,
                            'sign': sign
                        }, allow_redirects=False)
    jsonRaw = jsbeautifier.beautify(res.text)
    jsonObj = json.loads(jsonRaw)
except:
    login = True
    log.error("登录失败")
    log.response(jsonRaw, 1)

token = jsonObj['data']['token']
log.success("登录成功, token = " + token)
print()
timestamp = time.time()
username = jsonObj['data']['username']
sign = sha1(str(timestamp) + username + token)


dirname = os.path.dirname(__file__)
log.info('/upload/image  '+'上传一个png格式图片文件', 0, getNum())
with open(os.path.join(dirname, 'assets/google.png'), 'rb') as pngfile:
    res = requests.post(baseUrl + "/upload/image",
                        data={
                            'username': username,
                            'timestamp': timestamp,
                            'sign': sign,
                        },
                        files={'file': ('google.png', pngfile, 'image/png')}, allow_redirects=False)
jsonRaw = jsbeautifier.beautify(res.text)
##log.response(jsonRaw, 1)
jsonPass = False
try:
    jsonObj = json.loads(jsonRaw)
    jsonPass = True
except:
    log.error("响应 body 不是有效 json")
    log.response(jsonRaw, 1)

if jsonPass:
    log.info("ret: " + str(jsonObj['ret']), 1)
    if 'ret' in jsonObj and jsonObj['ret'] == 200:
        log.info("url: " + str(jsonObj['data']['url']), 1)
        log.success('测试通过\n', 1)

log.info('/upload/image  '+'上传一个php文件, 将mime伪造为正常png文件', 0, getNum())
with open(os.path.join(dirname, 'assets/test.php'), 'rb') as pngfile:
    res = requests.post(baseUrl + "/upload/image",
                        data={
                            'username': username,
                            'timestamp': timestamp,
                            'sign': sign,
                        },
                        files={'file': ('google.png', pngfile, 'image/png')}, allow_redirects=False)
jsonRaw = jsbeautifier.beautify(res.text)
# log.response(jsonRaw, 1)
jsonPass = False
try:
    jsonObj = json.loads(jsonRaw)
    jsonPass = True
except:
    log.error("响应 body 不是有效 json")

if jsonPass:
    log.info("ret: " + str(jsonObj['ret']), 1)
    if 'ret' in jsonObj and jsonObj['ret'] == 23333:
        log.info("msg: " + str(jsonObj['msg']), 1)
        log.success('测试通过\n', 1)

stopwatch.stop()
log.info("测试结束, 一共 " + str(testNum) + " 个测试, 用时 " + str(stopwatch))
print()
